File: hw/sanchez/u3/hw18OOPConceptsZoo/ZooPython/model/farm_animal.py
from datetime import datetime
from abc import ABC, abstractmethod

class FarmAnimal(ABC):
    def __init__(self, id: int, breed: str, born_on: datetime, weight: float):
        self.id = id
        self.breed = breed
        self.born_on = born_on
        self.weight = weight
    
    def get_age_in_months(self) -> int:
        now = datetime.now()
        diff = now - self.born_on
        months = diff.days // 30
        return months
    
    @abstractmethod
    def food(self, food):
        pass
    
    def __str__(self):
        return f"FarmAnimal(id={self.id}, breed='{self.breed}', born_on={self.born_on}, weight={self.weight})"

File: hw/sanchez/u3/hw18OOPConceptsZoo/OOPConceptsZoo/src/main/java/ec/edu/espe/oopconceptszoo/model/FarmAnimal.java
package ec.edu.espe.oopconceptszoo.model;

import java.util.Date;

/**
 * @author Joel Sanchez, The Softwarriors, @ESPE
 *
 */
public abstract class FarmAnimal {
    private int id;
    private String breed;
    private Date bornOn;
    private float weight;
    
    public FarmAnimal(int id, String breed, Date bornOn, float weight) {
        this.id = id;
        this.breed = breed;
        this.bornOn = bornOn;
        this.weight = weight;
    }
    
    public int getId() { return id; }
    public void setId(int id) { this.id = id; }
    public String getBreed() { return breed; }
    public void setBreed(String breed) { this.breed = breed; }
    public Date getBornOn() { return bornOn; }
    public void setBornOn(Date bornOn) { this.bornOn = bornOn; }
    public float getWeight() { return weight; }
    public void setWeight(float weight) { this.weight = weight; }
    
    public int getAgeInMonths() {
        Date now = new Date();
        long diff = now.getTime() - bornOn.getTime();
        long months = diff / (1000L * 60 * 60 * 24 * 30);
        return (int) months;
    }
    
    public abstract void food(Food food);
    
    @Override
    public String toString() {
        return "FarmAnimal{" +
                "id=" + id +
                ", breed='" + breed + '\'' +
                ", bornOn=" + bornOn +
                ", weight=" + weight +
                '}';
    }
}
